koku/forecast/forecast.py

```diff
@@ -15,88 +15,226 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #
 """Base forecasting module."""
-from datetime import datetime
+import logging
+from abc import ABC
+from abc import abstractmethod
 from datetime import timedelta
 
-from api.models import Provider
-
+import statsmodels.api as sm
+from statsmodels.sandbox.regression.predstd import wls_prediction_std
+from statsmodels.tools.sm_exceptions import ValueWarning
+from tenant_schemas.utils import tenant_context
 
-class Forecast:
+from api.models import Provider
+from api.query_filter import QueryFilterCollection
+from api.report.all.openshift.provider_map import OCPAllProviderMap
+from api.report.aws.openshift.provider_map import OCPAWSProviderMap
+from api.report.aws.provider_map import AWSProviderMap
+from api.report.azure.openshift.provider_map import OCPAzureProviderMap
+from api.report.azure.provider_map import AzureProviderMap
+from api.report.ocp.provider_map import OCPProviderMap
+from api.utils import DateHelper
+
+
+LOG = logging.getLogger(__name__)
+FAKE_RESPONSE = [
+    {
+        "date": (DateHelper().now + timedelta(days=1)).strftime("%Y-%m-%d"),
+        "value": 1,
+        "confidence_min": 0,
+        "confidence_max": 2,
+    },
+    {
+        "date": (DateHelper().now + timedelta(days=2)).strftime("%Y-%m-%d"),
+        "value": 2,
+        "confidence_min": 1,
+        "confidence_max": 3,
+    },
+    {
+        "date": (DateHelper().now + timedelta(days=3)).strftime("%Y-%m-%d"),
+        "value": 3,
+        "confidence_min": 2,
+        "confidence_max": 4,
+    },
+    {
+        "date": (DateHelper().now + timedelta(days=4)).strftime("%Y-%m-%d"),
+        "value": 4,
+        "confidence_min": 3,
+        "confidence_max": 5,
+    },
+    {
+        "date": (DateHelper().now + timedelta(days=5)).strftime("%Y-%m-%d"),
+        "value": 5,
+        "confidence_min": 4,
+        "confidence_max": 6,
+    },
+]
+
+
+class Forecast(ABC):
     """Base forecasting class."""
 
+    # the minimum number of data points needed to use the current month's data.
+    # if we have fewer than this many data points, fall back to using the previous month's data.
+    #
+    # this number is chosen in part because statsmodels.stats.stattools.omni_normtest() needs at least eight data
+    # points to test for normal distribution.
+    MINIMUM = 8
+    REPORT_TYPE = "costs"
+    PRECISION = 8
+    dh = DateHelper()
+
     def __init__(self, query_params):
         """Class Constructor."""
-        pass
+        self.params = query_params
+        self.cost_summary_table = self.provider_map(self.provider, self.REPORT_TYPE).views.get("costs").get("default")
+
+        time_scope_units = query_params.get_filter("time_scope_units", "month")
+        time_scope_value = int(query_params.get_filter("time_scope_value", -1))
```

Comment:
@dlabrecq will be calling the API with `filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=daily`

It looks like it will automatically grab last months data to build the forecast if there are less than 8 days of data in the current month.  Is this correct?

Reply:
@dccurtis this is correct. My goal is to maximize our chances of producing a high-quality forecast. If we can minimize or avoid cases where the data set is known to be limited, that's a good thing. 

With this mechanism, the only time we'll have less than ~1 week of data is the very first week the customer onboards a new data source. In that case, we'll be able to deliver naive forecasts within a customer's first 2-3 days, and providing fairly reasonable forecasts after their first week.

If you want to talk through the other approaches I considered and why this is what I settled on as the starting point, I'm open to discussing further. :slightly_smiling_face: